### ISSUE TITLE
[BUG]: compare_models() different results between 3.0.0 and >= 3.0.1

### ISSUE THREAD

**Issue by fgma** [pycaret/pycaret#3570]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [x] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

After updating pycaret to any version >= 3.0.1 I get unexpected results when running compare_models(). Input is classical tabular data. Expected result is a high score for rf/catboost/lightgbm/xgboost. After updating the result for these models is a lot worse and the best model ends up being knn.

Output running 3.0.0:

```
                                    Model  Accuracy     AUC  Recall   Prec.  \
catboost              CatBoost Classifier    0.9164  0.9531  0.5821  0.8957   
lightgbm  Light Gradient Boosting Machine    0.9117  0.9608  0.5982  0.8560   
xgboost         Extreme Gradient Boosting    0.9094  0.9419  0.6393  0.8143   
gbc          Gradient Boosting Classifier    0.9071  0.9348  0.6375  0.8231   
rf               Random Forest Classifier    0.9049  0.9552  0.5071  0.9314   
ada                  Ada Boost Classifier    0.8933  0.9225  0.5929  0.7366   
et                 Extra Trees Classifier    0.8932  0.9470  0.4143  0.9000   
knn                K Neighbors Classifier    0.8863  0.8659  0.3839  0.8583   
ridge                    Ridge Classifier    0.8863  0.0000  0.5446  0.7632   
lda          Linear Discriminant Analysis    0.8863  0.8883  0.5696  0.7524   
lr                    Logistic Regression    0.8817  0.8737  0.5661  0.7338   
svm                   SVM - Linear Kernel    0.8746  0.0000  0.5679  0.6690   
dt               Decision Tree Classifier    0.8726  0.7565  0.5804  0.6654   
nb                            Naive Bayes    0.8539  0.8596  0.5036  0.7028   
qda       Quadratic Discriminant Analysis    0.8422  0.6211  0.1036  0.5500   
dummy                    Dummy Classifier    0.8260  0.5000  0.0000  0.0000   

              F1   Kappa     MCC  TT (Sec)  
catboost  0.6927  0.6487  0.6750     0.089  
lightgbm  0.6896  0.6419  0.6638     0.055  
xgboost   0.7091  0.6567  0.6680     0.058  
gbc       0.7009  0.6482  0.6657     0.057  
rf        0.6341  0.5876  0.6345     0.068  
ada       0.6506  0.5889  0.5975     0.057  
et        0.5523  0.5068  0.5610     0.062  
knn       0.5140  0.4672  0.5214     0.210  
ridge     0.6200  0.5565  0.5753     0.050  
lda       0.6313  0.5671  0.5832     0.050  
lr        0.6173  0.5507  0.5685     0.281  
svm       0.6094  0.5360  0.5414     0.058  
dt        0.6071  0.5324  0.5414     0.063  
nb        0.5561  0.4734  0.4991     0.061  
qda       0.1717  0.1462  0.2122     0.051  
dummy     0.0000  0.0000  0.0000     0.050  

```

Output after updating to >= 3.0.1:

```
                                    Model  Accuracy     AUC  Recall   Prec.  \
knn                K Neighbors Classifier    0.8724  0.8045  0.2768  0.8000   
svm                   SVM - Linear Kernel    0.8375  0.0000  0.1036  0.3300   
et                 Extra Trees Classifier    0.8353  0.9168  0.0500  0.3000   
lr                    Logistic Regression    0.8352  0.8951  0.0768  0.4000   
nb                            Naive Bayes    0.8260  0.5000  0.0000  0.0000   
dt               Decision Tree Classifier    0.8260  0.5000  0.0000  0.0000   
ridge                    Ridge Classifier    0.8260  0.0000  0.0000  0.0000   
rf               Random Forest Classifier    0.8260  0.9487  0.0000  0.0000   
qda       Quadratic Discriminant Analysis    0.8260  0.5000  0.0000  0.0000   
ada                  Ada Boost Classifier    0.8260  0.5000  0.0000  0.0000   
gbc          Gradient Boosting Classifier    0.8260  0.5000  0.0000  0.0000   
lda          Linear Discriminant Analysis    0.8260  0.5000  0.0000  0.0000   
xgboost         Extreme Gradient Boosting    0.8260  0.8185  0.0000  0.0000   
catboost              CatBoost Classifier    0.8260  0.8162  0.0000  0.0000   
dummy                    Dummy Classifier    0.8260  0.5000  0.0000  0.0000   
lightgbm  Light Gradient Boosting Machine    0.6376  0.6651  0.3000  0.0558   

              F1   Kappa     MCC  TT (Sec)  
knn       0.3932  0.3550  0.4220     0.194  
svm       0.1488  0.1241  0.1514     0.044  
et        0.0844  0.0729  0.1107     0.059  
lr        0.1272  0.1034  0.1500     0.262  
nb        0.0000  0.0000  0.0000     0.047  
dt        0.0000  0.0000  0.0000     0.047  
ridge     0.0000  0.0000  0.0000     0.044  
rf        0.0000  0.0000  0.0000     0.061  
qda       0.0000  0.0000  0.0000     0.048  
ada       0.0000  0.0000  0.0000     0.049  
gbc       0.0000  0.0000  0.0000     0.054  
lda       0.0000  0.0000  0.0000     0.050  
xgboost   0.0000  0.0000  0.0000     0.053  
catboost  0.0000  0.0000  0.0000     0.083  
dummy     0.0000  0.0000  0.0000     0.047  
lightgbm  0.0941  0.0000  0.0000     0.053  

```

### Reproducible Example

```python
pycaret_setup = pycaret.classification.setup(data_full, target='target', normalize=True, remove_multicollinearity=False, polynomial_features=False, session_id=123)
pycaret_best = pycaret_setup.compare_models()
pycaret_results = pycaret_setup.pull()
print(pycaret_results)
```

### Expected Behavior

Results of compare_models() should have a similar score.

### Actual Results

```python-traceback
Model scores are a lot worse using the new version.
```

### Installed Versions

<details>

System:
    python: 3.10.11 | packaged by Anaconda, Inc. | (main, Apr 20 2023, 18:56:50) [MSC v.1916 64 bit (AMD64)]
executable: c:\Users\user\anaconda3\envs\ml\python.exe
   machine: Windows-10-10.0.22621-SP0

PyCaret required dependencies:
                 pip: 23.0.1
          setuptools: 66.0.0
             pycaret: 3.0.1
             IPython: 8.13.2
          ipywidgets: 8.0.6
                tqdm: 4.65.0
               numpy: 1.23.5
              pandas: 1.5.3
              jinja2: 3.1.2
               scipy: 1.9.3
              joblib: 1.2.0
             sklearn: 1.2.2
                pyod: 1.0.9
            imblearn: 0.10.1
   category_encoders: 2.6.1
            lightgbm: 3.3.5
               numba: 0.57.0
            requests: 2.31.0
          matplotlib: 3.6.3
          scikitplot: 0.3.7
         yellowbrick: 1.5
              plotly: 5.14.1
             kaleido: 0.2.1
         statsmodels: 0.14.0
              sktime: 0.17.0
               tbats: 1.1.3
            pmdarima: 2.0.3
              psutil: 5.9.5

PyCaret optional dependencies:
                shap: Not installed
           interpret: Not installed
                umap: Not installed
    pandas_profiling: 4.2.0
  explainerdashboard: Not installed
             autoviz: Not installed
           fairlearn: Not installed
             xgboost: 1.7.5
            catboost: 1.2
              kmodes: Not installed
             mlxtend: Not installed
       statsforecast: Not installed
        tune_sklearn: Not installed
                 ray: Not installed
            hyperopt: Not installed
              optuna: Not installed
               skopt: Not installed
              mlflow: Not installed
              gradio: Not installed
             fastapi: Not installed
             uvicorn: Not installed
              m2cgen: Not installed
           evidently: Not installed
               fugue: Not installed
           streamlit: Not installed
             prophet: Not installed

</details>

**Comment by fgma** [comment]
I've reproduced similar behaviour using the titanic dataset. Overall model accuracy is a lot worse in 3.0.1.

```
import pycaret.classification
import pycaret.datasets

data_full = pycaret.datasets.get_data('titanic')

pycaret_setup = pycaret.classification.setup(data_full, target='Survived', normalize=True, remove_multicollinearity=False, polynomial_features=False, session_id=123)
pycaret_best = pycaret_setup.compare_models()
pycaret_results = pycaret_setup.pull()
print(pycaret_results)
```

Output on 3.0.0:

```
                                    Model  Accuracy     AUC  Recall   Prec.  \
lr                    Logistic Regression    0.8040  0.8677  0.6904  0.7726   
knn                K Neighbors Classifier    0.8040  0.8287  0.6862  0.7771   
ridge                    Ridge Classifier    0.8040  0.0000  0.6987  0.7721   
lda          Linear Discriminant Analysis    0.8040  0.8676  0.6987  0.7721   
et                 Extra Trees Classifier    0.7978  0.8321  0.5775  0.8466   
svm                   SVM - Linear Kernel    0.7528  0.0000  0.6196  0.7098   
nb                            Naive Bayes    0.7449  0.8233  0.4899  0.7627   
rf               Random Forest Classifier    0.7126  0.8299  0.3054  0.8674   
gbc          Gradient Boosting Classifier    0.7046  0.7553  0.2850  0.8593   
lightgbm  Light Gradient Boosting Machine    0.7014  0.7389  0.2931  0.8215   
catboost              CatBoost Classifier    0.7014  0.8312  0.2888  0.8290   
ada                  Ada Boost Classifier    0.6966  0.8033  0.2808  0.8197   
xgboost         Extreme Gradient Boosting    0.6966  0.8018  0.2804  0.8165   
dt               Decision Tree Classifier    0.6934  0.6132  0.2679  0.8127   
dummy                    Dummy Classifier    0.6164  0.5000  0.0000  0.0000   
qda       Quadratic Discriminant Analysis    0.5566  0.5968  0.5750  0.4884   

              F1   Kappa     MCC  TT (Sec)  
lr        0.7258  0.5751  0.5798     0.311  
knn       0.7236  0.5742  0.5806     0.247  
ridge     0.7306  0.5777  0.5819     0.082  
lda       0.7306  0.5777  0.5819     0.087  
et        0.6826  0.5437  0.5669     0.102  
svm       0.6567  0.4659  0.4723     0.085  
nb        0.5845  0.4183  0.4449     0.090  
rf        0.4422  0.3075  0.3855     0.102  
gbc       0.4177  0.2854  0.3644     0.098  
lightgbm  0.4228  0.2815  0.3507     0.093  
catboost  0.4191  0.2800  0.3517     0.126  
ada       0.4087  0.2684  0.3398     0.104  
xgboost   0.4084  0.2680  0.3384     0.094  
dt        0.3959  0.2578  0.3296     0.090  
dummy     0.0000  0.0000  0.0000     0.087  
qda       0.4348  0.1154  0.1279     0.096  

```


Output on 3.0.1:

```
                                    Model  Accuracy     AUC  Recall   Prec.  \
knn                K Neighbors Classifier    0.7542  0.7993  0.4815  0.7996   
et                 Extra Trees Classifier    0.7400  0.7837  0.4774  0.7654   
lr                    Logistic Regression    0.6853  0.8208  0.2679  0.7670   
svm                   SVM - Linear Kernel    0.6612  0.0000  0.4011  0.7339   
nb                            Naive Bayes    0.6164  0.5000  0.0000  0.0000   
dt               Decision Tree Classifier    0.6164  0.5000  0.0000  0.0000   
ridge                    Ridge Classifier    0.6164  0.0000  0.0000  0.0000   
rf               Random Forest Classifier    0.6164  0.7890  0.0000  0.0000   
ada                  Ada Boost Classifier    0.6164  0.5000  0.0000  0.0000   
gbc          Gradient Boosting Classifier    0.6164  0.5000  0.0000  0.0000   
lda          Linear Discriminant Analysis    0.6164  0.5000  0.0000  0.0000   
xgboost         Extreme Gradient Boosting    0.6164  0.6404  0.0000  0.0000   
lightgbm  Light Gradient Boosting Machine    0.6164  0.5200  0.0000  0.0000   
catboost              CatBoost Classifier    0.6164  0.7567  0.0000  0.0000   
dummy                    Dummy Classifier    0.6164  0.5000  0.0000  0.0000   
qda       Quadratic Discriminant Analysis    0.5204  0.5000  0.4000  0.1520   

              F1   Kappa     MCC  TT (Sec)  
knn       0.5947  0.4367  0.4680     0.277  
et        0.5815  0.4088  0.4356     0.179  
lr        0.3922  0.2423  0.3052     0.335  
svm       0.4374  0.2379  0.2910     0.093  
nb        0.0000  0.0000  0.0000     0.103  
dt        0.0000  0.0000  0.0000     0.107  
ridge     0.0000  0.0000  0.0000     0.093  
rf        0.0000  0.0000  0.0000     0.118  
ada       0.0000  0.0000  0.0000     0.097  
gbc       0.0000  0.0000  0.0000     0.103  
lda       0.0000  0.0000  0.0000     0.110  
xgboost   0.0000  0.0000  0.0000     0.108  
lightgbm  0.0000  0.0000  0.0000     0.107  
catboost  0.0000  0.0000  0.0000     0.145  
dummy     0.0000  0.0000  0.0000     0.100  
qda       0.2203  0.0000  0.0000     0.101  
```

**Comment by celestinoxp** [comment]
as far as i know it looks like the problem is in 3.0.0 but it was resolved in 3.0.1.
Please, can you check in 3.0.2, if it is equal to the values of 3.0.1 ?
If it's solved, you can close this issue...

**Comment by celestinoxp** [comment]
@Yard1 might have some information

**Comment by tvdboom** [comment]
Hi @fgma. First of all, if I rerun your titanic code on the current master branch I get way better results than you are showing:

![image](https://github.com/pycaret/pycaret/assets/32366550/0aeda658-28d6-4029-9928-6c3ff14f8bb4)

But you are right that the results may change since v3.0.1. The reason is we changed the default encoding method for high cardinality features from LeaveOneOut to Target encoding. The reason being explainer in #3507

To check this is indeed the case, try using LeaveOneOut as encoding method and you should get the same results as in 3.0.0

```python
import pycaret.classification
import pycaret.datasets
from category_encoders.leave_one_out import LeaveOneOutEncoder

data_full = pycaret.datasets.get_data('titanic')

pycaret_setup = pycaret.classification.setup(data_full, target='Survived', normalize=True, remove_multicollinearity=False, polynomial_features=False, encoding_method=LeaveOneOutEncoder(), session_id=123)
pycaret_best = pycaret_setup.compare_models()
pycaret_results = pycaret_setup.pull()
print(pycaret_results)
```
